libraries/utils/src/render-lines-on-sp.ts

```diff
@@ -63,58 +70,66 @@ export class ScatterPlotLinesRenderer {
+            const pointTo = this.sp.worldToScreen(this.xAxisCol.get(pointToIdx), this.yAxisCol.get(pointToIdx));
+            const bX = pointTo?.x;
+            const bY = pointTo?.y;
+            if (aX && aY && bX && bY) {
+                const line = new Path2D();
+                this.paths[i] = line;
+                line.moveTo(aX!, aY!);
+                const color = this.lines.color ?? this.lines.colors?.[i] ? this.lines.colors?.[i] : '0,128,0';
+                const opacity = this.lines.opacity ?? this.lines.opacities?.[i] ? this.lines.opacities?.[i] : 1;
+                this.ctx.strokeStyle = `rgba(${color},${opacity})`;
+                this.ctx.lineWidth = this.lines.width ? this.lines.width : this.lines.widths?.[i] ? i === this.currentLineIdx ? this.lines.widths?.[i] + 2 :
                     this.lines.widths?.[i] : i === this.currentLineIdx ? 3 : 1;
-                    let midPoint: DG.Point = this.midPoint(pointFrom, pointTo);
-                    if (this.numberOfLinesPerPair[pairStr] > 1 || linesPerPairCounter[pairStr] > 1) {
-                        midPoint = this.findPerpendicularPointOnCurve(linesPerPairCounter[pairStr],
-                            pointFrom.x, pointFrom.y, midPoint.x, midPoint.y)
-                    }
-                    line.quadraticCurveTo(midPoint.x, midPoint.y, bX, bY);
-                    if (this.lines.drawArrows?.getBit(i))
-                        this.canvasArrow(line, bX, bY, midPoint.x, midPoint.y);
-                    this.ctx.beginPath();
-                    this.ctx.stroke(line);
-                    this.ctx.closePath();
+                let midPoint: DG.Point = this.midPoint(pointFrom, pointTo);
+                if (this.multipleLinesIndices[i]) {
+                    midPoint = this.findPerpendicularPointOnCurve(this.multipleLinesIndices[i],
+                        pointFrom.x, pointFrom.y, midPoint.x, midPoint.y)
                 }
+                line.quadraticCurveTo(midPoint.x, midPoint.y, bX, bY);
+                if (this.lines.drawArrows ?? this.lines.drawArrowsArr?.getBit(i))
+                    this.canvasArrow(line, changedPoints ? aX : bX, changedPoints ? aY : bY, midPoint.x, midPoint.y);
+                this.ctx.beginPath();
+                this.ctx.stroke(line);
+                this.ctx.closePath();
             }
         }
     }
 
-    countLinesPerPair(): void {
-        for (let i = 0; i < this.lines.from.length; i++) {
-            const pairStr = this.getStringPair(this.lines.from[i], this.lines.to[i]);
-            !this.numberOfLinesPerPair[pairStr] ? this.numberOfLinesPerPair[pairStr] = 1 : this.numberOfLinesPerPair[pairStr]++;
-        }
-    }
-
-    getStringPair(point1: number, point2: number): string {
-        let p1 = point1;
-        let p2 = point2;
-        if (point1 > point2) {
-            p1 = point2;
-            p2 = point1;
+    createMultiLinesIndices(): void {
+        const arrayIdxs = new Array<number>(this.lines.from.length);
```

Comment:
Why are you using an Array and not a strongly-type, super-efficient Int32List?

Reply:
Changed to BitArray